README
-- ===========================================================================
      Revision: $Revision: 1166 $
          Date: $Date: 2016-05-10 09:21:03 +0200 (Wt, 10 maj 2016) $
   Last change: $Author: cnitsa $
-- ===========================================================================

TimeSlotTracker (TST) is a program for registering time spent on some work.
The main rule is to provide a flexible way to construct a task tree composed
of projects, (sub)projects and simple task inside that projects.
Time is registered in slots, it means if you had worked more then once per day
on some task you will have several slots registered, not only summarized time.

Project is founded by Bartlomiej Marciniak (qqbart),
then joined Zbigniew Ogledzki (zgibek)
Then we are very pleased that fellunia (Felicja Suska) joined to us!
Then joined Andrey Glazachev (cnitsa).


Main points:
-- ---------------
* Files should be saved in UTF-8 format
* Files should use unix end of line markers
* Files should use spaces (ident = 2)
* Files should contains svn's tags for easy recognizing them (Revision, Date, Author)
* Source files should be well documented to generate javadoc
* Whole project will be compiled using ant.  Ant have several targets, to show run "ant -p"
* And requires build.properties file with following params:
** username=<your sf name>, userpass=<your sf password>, keystorepass=<key store password>, keypass=<key password>
* Needed JRE 1.6+
* Exist eclipse/jetbrains idea project files


Main ideas:
-- ---------------
* User can register some task and subtask
* Every task has properties:
  - name
  - description (can be even a large one, maybe in html format in the future)
* Task are realized in time slots:
  - start & finish datetime 
  - description if given
* Time is summarized up in tree up to main branch
* Only one task can be timed in one time. 
  If one task is active and somebody starts another one the previous active
  should be finished and then the new one started.
* Data should be keep in xml format
* In the future we should use some kind of transaction log
* Time slots can be manually edited to correct (or just enter) values
* Data should be saved in configurable amount of time (for example 5 minutes).
  At the beginning the whole xml file should be regenerated. 
  In the future only transaction log should be written immediately and 
  the xml file should be written every one hour (for example)
* To detect inactive time TT should ask a question every X minutes
  (for example 15 minutes). If break is detected user have to decided what
  to do with break time.
* TST should be a localized friendly software, so every message should be got
  from properties file.


Roadmap:
-- ---------------
* see ChangeLog file


ChangeLog
============================== version 1.3.1 ==============================

2016-05-26  Glazachev Andrey <cnitsa@users.sourceforge.net>
	* Change: localizations for added memory/data save status field (toolbar)
	* Change: toolbar: pause/stop actions does not depend on selected task

2016-05-11  Glazachev Andrey <cnitsa@users.sourceforge.net>
	* New: added memory/data save status field (toolbar)

2016-05-10  Glazachev Andrey <cnitsa@users.sourceforge.net>
	* Fix: rss url updated (used for version update)

============================== version 1.3.0 ==============================

2016-05-06  Glazachev Andrey <cnitsa@users.sourceforge.net>
	* New: added jira6/7 support for add worklog method

2016-04-09  Glazachev Andrey <cnitsa@users.sourceforge.net>
	* Change: jira version introduced

2016-01-16  Glazachev Andrey <cnitsa@users.sourceforge.net>
	* Change: debian deps: added openjdk-8-jre

2016-01-16  Glazachev Andrey <cnitsa@users.sourceforge.net>
	* New: DaysTree/monthNode: show delta current-planned

2016-01-07  Glazachev Andrey <cnitsa@users.sourceforge.net>
	* Fix: #70 UAC for standard user not expected

2016-01-03  Glazachev Andrey <cnitsa@users.sourceforge.net>
	* Fix: debian package compatibility updates

2016-01-02  Glazachev Andrey <cnitsa@users.sourceforge.net>
	* Fix: Bug#80 Help > Check Updates > Error getting last version from sf.net

2016-01-01  Glazachev Andrey <cnitsa@users.sourceforge.net>
	* Patch: #83 Always re-assigning new timeslot ID on XML load

2015-05-09  Glazachev Andrey <cnitsa@users.sourceforge.net>
	* Fix: Bug#81 Report Generation Error

2015-04-11  Glazachev Andrey <cnitsa@users.sourceforge.net>
	* Change: RFE#93 Italian Translation (thanks to Giuseppe Morabito)

2015-01-05  Glazachev Andrey <cnitsa@users.sourceforge.net>
	* Fix: Bug#79 "My Tasks" Main Node - Disappearing Sub-Tasks

============================== version 1.2.9 ==============================

2014-08-30  Glazachev Andrey <cnitsa@users.sourceforge.net>
	* Change: RFE#92 Dependencies need to be updated (debian installer)

2014-05-31  Glazachev Andrey <cnitsa@users.sourceforge.net>
	* Change: documentation url changes: to https://sourceforge.net/p/timeslottracker/wiki/Documentation/ (drop Trac support)

2014-05-02  Glazachev Andrey <cnitsa@users.sourceforge.net>
	* Change: german translation updated (thanks to Jutta Albrecht)

2014-04-25  Glazachev Andrey <cnitsa@users.sourceforge.net>
	* Change: restart timing action: simplified confidion's logic
	* Change: restart timing action: user now can cancel restart timing if dialog (with time) occurs
	* Change: start/pause/restart timing refactored: some regress from 2014-04-04, 2014-03-31 commits

2014-04-05  Glazachev Andrey <cnitsa@users.sourceforge.net>
	* Change: added missing russian localization strings

2014-04-04  Gerth Hermansson <ghermans@users.sourceforge.net>
	* Change: Added in, User idle detection configuration, for if "Show dialog to set a custom time for restart of task after idle".
	* Change: TimeSlotTracker_en.properties added label for configuration.
	* Change: Configuration.java added constant for label
	* Change: UserIdleDetectorConfigTab.java added checkbox for "Show dialog to set a custom time for restart of task after idle".
	* Change: UserIdleDetector.java updated actionPerformed() method.
	* Change: Starter.java modified method restartTiming to check if user has selected
	          to have "Show dialog to set a custom time for restart of task after idle" shown
	          in addition to the previously implemented "Show dialog to set a custom time for restart of task".
	* Change: TimeSlotsInterface.java added new method getSelected() to be able to get the description of a non-active but selected
	          TimeSlot (patch after noticing that restart timing from toolbar worked differently than restart timing from context menu).
	* Change: Timeslots.java implemented new method getSelected().

2014-03-31  Gerth Hermansson <ghermans@users.sourceforge.net>
	* Change: Added customization for if "Show task has just started" message will be shown.
	* Change: TimeSlotTracker_en.properties added label for configuration.
	* Change: Configuration.java added constant for label
	* Change: ConfirmationsTab.java added checkbox for "Show task has just started" message.
	* Change: Starter.java modified method startTiming to check if user has selected
	*         to have "Show task has just started" message to be shown.

2014-03-30  Glazachev Andrey <cnitsa@users.sourceforge.net>
	* New: RFE#88 Cloning of subtrees: refactoring: updating task model
	* Change: TimeSlotRestartDialog remove unused logic
	* Change: TimeSlotRestartDialog/TimeSlotInputDialog: start time field name now ending with : symbol
	* Change: updated jgoodies looks to 2.6.0, jgoodies common to 1.8.0

2014-03-28  Gerth Hermansson <ghermans@users.sourceforge.net>
	* New: RFE#88 Cloning of subtrees
	* New: Create new timeslot now have start date/time field
	* New: New dialog for restart timing

2014-03-24  Glazachev Andrey <cnitsa@users.sourceforge.net>
	* New: RFE#91 Make "Create new time slot" dialog aware of hidden tasks

2014-03-22  Glazachev Andrey <cnitsa@users.sourceforge.net>
	* Patch: Html: Project Tasks report (thanks to Gerth Hermansson)
	* New: Added missing Mac OS X's icons

2014-03-16  Glazachev Andrey <cnitsa@users.sourceforge.net>
	* Patch: Mac OS X's monochrome icons in System Tray (thanks to Gerth Hermansson)

2014-03-09  Glazachev Andrey <cnitsa@users.sourceforge.net>
	* Patch: Weeks not sorted on week number" in "Task by days" view (thanks to Gerth Hermansson)

2014-03-01  Glazachev Andrey <cnitsa@users.sourceforge.net>
	* Fix: Bug#72 Popup current activity option not working

2014-02-16  Glazachev Andrey <cnitsa@users.sourceforge.net>
	* New: RFE#38 Purge old data, #90 Archiving functionality  - removing timeslots before given date 

2014-01-26  Glazachev Andrey <cnitsa@users.sourceforge.net>

resources/java/net/sf/timeslottracker/gui/TimeSlotInputDialog.java
package net.sf.timeslottracker.gui;

import java.awt.FlowLayout;
import java.awt.GridBagConstraints;
import java.awt.HeadlessException;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.awt.event.InputEvent;
import java.awt.event.ItemEvent;
import java.awt.event.ItemListener;
import java.awt.event.KeyEvent;
import java.awt.event.WindowAdapter;
import java.awt.event.WindowEvent;
import java.util.Calendar;
import java.util.Date;

import javax.swing.AbstractAction;
import javax.swing.JButton;
import javax.swing.JComboBox;
import javax.swing.JComponent;
import javax.swing.JPanel;
import javax.swing.KeyStroke;

import net.sf.timeslottracker.data.Task;
import net.sf.timeslottracker.gui.taskmodel.TaskModel;
import net.sf.timeslottracker.gui.taskmodel.TaskModelFactory;
import net.sf.timeslottracker.gui.taskmodel.TaskValue;
import net.sf.timeslottracker.utils.SwingUtils;

/**
 * TimeSlot create dialog
 * <p>
 * It contains input field for description with some last descriptions from
 * active task (history)
 * <p>
 * Also contains task selection. After updated description history.
 * 
 * @version File version: $Revision: 1118 $, $Date: 2009-08-04 19:26:06 +0700
 *          (Tue, 04 Aug 2009) $
 * @author Last change: $Author: cnitsa $
 */
@SuppressWarnings("serial")
public class TimeSlotInputDialog extends AbstractSimplePanelDialog {

  private final AbstractAction ACTION_CANCEL = new AbstractAction() {
    public void actionPerformed(ActionEvent e) {
      wasCancel = true;
      close();
    }
  };

  private final ActionListener ACTION_NEW_ISSUE_TASK = new AbstractAction() {
    public void actionPerformed(ActionEvent e) {
      changeActiveTask();

      layoutManager.getTasksInterface().addTaskFromIssueTracker();
      updateTaskModel();

      selectActiveTask();
    }
  };

  private final ActionListener ACTION_NEW_TASK = new AbstractAction() {
    public void actionPerformed(ActionEvent e) {
      changeActiveTask();

      layoutManager.getTasksInterface().add(null, null);
      updateTaskModel();

      selectActiveTask();
    }
  };

  private final AbstractAction ACTION_OK = new AbstractAction("OK") {
    public void actionPerformed(ActionEvent e) {
      wasCancel = false;
      changeActiveTask();
      close();
    }
  };

  private DescriptionInputComboBox inputComboBox;

  private final LayoutManager layoutManager;

  private TaskModel taskmodel;

  private boolean wasCancel;

  private DatetimeEditPanel startDate;

  public TimeSlotInputDialog(LayoutManager layoutManager)
      throws HeadlessException {
    super(layoutManager, layoutManager
        .getCoreString("timing.start.input.title"));

    this.layoutManager = layoutManager;

    initActions();
  }

  private void initActions() {
    // connect cancelAction with ESC key
    getRootPane().registerKeyboardAction(ACTION_CANCEL,
        KeyStroke.getKeyStroke(KeyEvent.VK_ESCAPE, 0),
        JComponent.WHEN_IN_FOCUSED_WINDOW);

    // connect okAction with Enter key
    getRootPane().registerKeyboardAction(ACTION_OK,
        KeyStroke.getKeyStroke(KeyEvent.VK_ENTER, 0),
        JComponent.WHEN_IN_FOCUSED_WINDOW);

    // connect new task with ins key
    getRootPane().registerKeyboardAction(ACTION_NEW_TASK,
        KeyStroke.getKeyStroke(KeyEvent.VK_INSERT, 0),
        JComponent.WHEN_IN_FOCUSED_WINDOW);

    // connect new issue task with ins + shift key
    getRootPane().registerKeyboardAction(ACTION_NEW_ISSUE_TASK,
        KeyStroke.getKeyStroke(KeyEvent.VK_INSERT, InputEvent.ALT_MASK),
        JComponent.WHEN_IN_FOCUSED_WINDOW);

    addWindowListener(new WindowAdapter() {
      public void windowClosing(WindowEvent e) {
        ACTION_CANCEL.actionPerformed(new ActionEvent(this, 1, ""));
      }
    });
  }

  /**
   * @return timeslot description, may be null
   */
  public String getDescription() {
    return wasCancel ? null : inputComboBox.getDescription();
  }

  public Task getSelectedTask() {
    TaskValue selectedItem = (TaskValue) taskmodel.getSelectedItem();
    Task selectedTask = layoutManager.getTimeSlotTracker().getDataSource()
        .getTask(selectedItem.getId());
    return selectedTask;
  }

  private void changeActiveTask() {
    Task task = getSelectedTask();
    if (task != null) {
      layoutManager.getTasksInterface().selectTask(task);
    }
  }

  private void close() {
    SwingUtils.saveLocation(this);
    dispose();
  }

  protected JPanel createButtons() {
    FlowLayout layout = new FlowLayout();
    layout.setHgap(15);

    JButton saveButton = new JButton(
        layoutManager.getCoreString("timing.start.input.button.save"),
        layoutManager.getIcon("save"));
    saveButton.addActionListener(ACTION_OK);
    getRootPane().setDefaultButton(saveButton);

    JButton cancelButton = new JButton(
        layoutManager.getCoreString("timing.start.input.button.cancel"),
        layoutManager.getIcon("cancel"));
    cancelButton.addActionListener(ACTION_CANCEL);

    JPanel buttonsPanel = new JPanel(layout);
    buttonsPanel.add(cancelButton);
    buttonsPanel.add(saveButton);

    return buttonsPanel;
  }

  @Override
  protected DialogPanel getDialogPanel() {
    return new DialogPanel(GridBagConstraints.HORIZONTAL, 0.0);
  }

  @Override
  protected void fillDialogPanel(DialogPanel panel) {
    // manipulation with task
    taskmodel = createTaskModel();
    selectActiveTask();
    Task selectedTask = getSelectedTask();

    // adding description
    inputComboBox = new DescriptionInputComboBox(layoutManager, false,
        selectedTask);
    panel.addRow(layoutManager.getCoreString("timing.start.input.prompt"),
        inputComboBox);

    // adding user selected start time
    boolean readonly = true;
    startDate = new DatetimeEditPanel(layoutManager, readonly, true, true);
    startDate.setDatetime(Calendar.getInstance().getTime());
    panel.addRow(coreString("editDialog.timeslot.start.date.name") + ":",
        startDate);

    // adding task
    JComboBox taskComboBox = new JComboBox(taskmodel);
    panel.addRow(layoutManager.getCoreString("timing.start.input.prompt.task"),
        taskComboBox);
    taskComboBox.addItemListener(new ItemListener() {
      @Override
      public void itemStateChanged(ItemEvent e) {
        inputComboBox.setActiveTask(getSelectedTask(), false);
      }
    });

    panel.addRow(layoutManager.getCoreString("timing.start.input.information"));
  }

  @Override
  protected void beforeShow() {
    pack();
    setResizable(true);
  }

  @Override
  protected int getDefaultHeight() {
    return (int) getSize().getHeight();
  }

  @Override
  protected int getDefaultWidth() {
    return 450;
  }

  private TaskModel createTaskModel() {
    return new TaskModelFactory(layoutManager).createTaskModel();
  }

  private void selectActiveTask() {
    Task actualTask = layoutManager.getTimeSlotsInterface().getSelectedTask();
    if (actualTask != null) {
      taskmodel.setSelectedItem(new TaskValue(actualTask));
    }
  }

  private void updateTaskModel() {
    inputComboBox.setModel(createTaskModel());
  }

  /**
   * @return Date - user selected timeslot start date, may be null
   */
  public Date getStartDate() {
    return startDate.getDatetime();
  }
}
